src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/unuseddeductions/service/UnusedDeductionsCalculationService.kt

```diff
@@ -0,0 +1,74 @@
+) {
+
+  fun calculate(adjustments: List<AdjustmentServiceAdjustment>, offenderNo: String): Int? {
+    val prisoner = prisonService.getOffenderDetail(offenderNo)
+    val sourceData = prisonService.getPrisonApiSourceData(prisoner).copy(bookingAndSentenceAdjustments = useAdjustmentsFromAdjustmentsApi(adjustments))
+
+    val calculationUserInputs = CalculationUserInputs(useOffenceIndicators = true)
+
+    var validationMessages = validationService.validateBeforeCalculation(sourceData, calculationUserInputs)
+    if (validationMessages.isNotEmpty()) {
+      return null
+    }
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    validationMessages = validationService.validateBeforeCalculation(booking)
+    if (validationMessages.isNotEmpty()) {
+      return null
+    }
+
+    val result = calculationService.calculateReleaseDates(booking)
+    val releaseDateTypes = listOf(ReleaseDateType.CRD, ReleaseDateType.ARD, ReleaseDateType.MTD)
+    val calculationResult = result.second
+    var releaseDate = calculationResult.dates.filter { releaseDateTypes.contains(it.key) }.minOfOrNull { it.value }
+    val releaseDateSentence = result.first.getAllExtractableSentences().find { it.sentenceCalculation.releaseDate == releaseDate }
+    val unusedDeductions = if (releaseDateSentence != null) { // The release date will be a PRRD, no unused deductions.
+      val maxDeductions =
+        releaseDateSentence.sentenceCalculation.numberOfDaysToDeterminateReleaseDate
+
+      val remand = adjustments.filter { it.adjustmentType == AdjustmentServiceAdjustmentType.REMAND }.map { it.daysBetween!! }.reduceOrNull { acc, it -> acc + it } ?: 0
+      val taggedBail = adjustments.filter { it.adjustmentType == AdjustmentServiceAdjustmentType.TAGGED_BAIL }.map { it.days!! }.reduceOrNull { acc, it -> acc + it } ?: 0
+      val deductions = taggedBail + remand
+      max(0, deductions - maxDeductions)
+    } else { 0 }
+    return unusedDeductions
+  }
+
+  private fun mapToBookingAdjustmentType(adjustmentType: AdjustmentServiceAdjustmentType): BookingAdjustmentType? {
+    return when (adjustmentType) {
+      AdjustmentServiceAdjustmentType.UNLAWFULLY_AT_LARGE -> BookingAdjustmentType.UNLAWFULLY_AT_LARGE
+      AdjustmentServiceAdjustmentType.RESTORATION_OF_ADDITIONAL_DAYS_AWARDED -> BookingAdjustmentType.RESTORED_ADDITIONAL_DAYS_AWARDED
+      AdjustmentServiceAdjustmentType.ADDITIONAL_DAYS_AWARDED -> BookingAdjustmentType.ADDITIONAL_DAYS_AWARDED
+      else -> null
+    }
+  }
+
+  private fun useAdjustmentsFromAdjustmentsApi(adjustments: List<AdjustmentServiceAdjustment>): BookingAndSentenceAdjustments {
+    return BookingAndSentenceAdjustments(
+      bookingAdjustments = adjustments.filter { mapToBookingAdjustmentType(it.adjustmentType) != null }
+        .map { BookingAdjustment(active = true, fromDate = it.fromDate!!, toDate = it.toDate, numberOfDays = it.effectiveDays!!, type = mapToBookingAdjustmentType(it.adjustmentType)!!) },
```

Comment:
Similar question to above here. If in reality the API always populates these properties, you could consider making them non nullable in the data class. Appreciate I don't know this API or the context as yet! :)

Reply:
Yea I think we may want to improve this. Currently most things in the DTO are nullable because the record could've been created in NOMIS where none of these fields are required. We could maybe have two different DTOs depending on the source of the record.